File: Packages/common/Source/Commands/UnlinkNode.ts
import {GetAsync, Command, AssertV, CommandMeta} from "web-vcore/nm/mobx-graphlink.js";
import {CE} from "web-vcore/nm/js-vextensions.js";
import {MapEdit, UserEdit} from "../CommandMacros.js";
import {GetNode, IsRootNode} from "../DB/nodes.js";
import {GetNodeL2} from "../DB/nodes/$node.js";
import {IsUserCreatorOrMod} from "../DB/users/$user.js";
import {NodeChildLink} from "../DB/nodeChildLinks/@NodeChildLink.js";
import {GetNodeChildLinks} from "../DB/nodeChildLinks.js";

// todo: add full-fledged checking to ensure that nodes are never orphaned by move commands (probably use parents recursion to find at least one map root)

@MapEdit
@UserEdit
@CommandMeta({
	payloadSchema: ()=>({}),
})
export class UnlinkNode extends Command<{mapID: string|n, parentID: string, childID: string}, {}> {
	allowOrphaning = false; // could also be named "asPartOfCut", to be consistent with ForUnlink_GetError parameter

	parentToChildLinks: NodeChildLink[];
	Validate() {
		const {parentID, childID} = this.payload;
		const childParents = GetNodeChildLinks(undefined, childID);
		this.parentToChildLinks = GetNodeChildLinks(parentID, childID);
		AssertV(this.parentToChildLinks.length <= 1, "There should not be more than 1 link between parent and child.");

		/* let {parentID, childID} = this.payload;
		let childNode = await GetNodeAsync(childID);
		let parentNodes = await GetNodeParentsAsync(childNode);
		Assert(parentNodes.length > 1, "Cannot unlink this child, as doing so would orphan it. Try deleting it instead."); */
		const oldData = GetNodeL2(childID);
		AssertV(oldData, "oldData was null.");

		const baseText = `Cannot unlink node #${oldData.id}, since `;
		AssertV(IsUserCreatorOrMod(this.userInfo.id, oldData), `${baseText}you are not its owner. (or a mod)`);
		AssertV(this.allowOrphaning || childParents.length > 1, `${baseText}doing so would orphan it. Try deleting it instead.`);
		AssertV(!IsRootNode(oldData), `${baseText}it's the root-node of a map.`);
		//AssertV(!IsNodeSubnode(oldData), `${baseText}it's a subnode. Try deleting it instead.`);
	}

	DeclareDBUpdates(db) {
		const {parentID, childID} = this.payload;
		db.set(`nodes/${childID}/.parents/.${parentID}`, null);
		db.set(`nodes/${parentID}/.children/.${childID}`, null);
		if (this.parentToChildLinks.length) {
			//db.set(`nodes/${parentID}/.childrenOrder`, CE(CE(this.parent_oldChildrenOrder).Except(childID)).IfEmptyThen(null));
			db.set(`nodeChildLinks/${this.parentToChildLinks[0].id}`, null);
		}
	}
}

File: Packages/app-server-rs/src/db/commands/unlink_node.rs
use rust_shared::async_graphql::{ID, SimpleObject, InputObject};
use rust_shared::rust_macros::wrap_slow_macros;
use rust_shared::serde_json::{Value, json};
use rust_shared::utils::db_constants::SYSTEM_USER_ID;
use rust_shared::{async_graphql, serde_json, anyhow, GQLError};
use rust_shared::async_graphql::{Object};
use rust_shared::utils::type_aliases::JSONValue;
use rust_shared::anyhow::{anyhow, Error, ensure};
use rust_shared::utils::time::{time_since_epoch_ms_i64};
use rust_shared::serde::{Deserialize};
use tracing::info;

use crate::db::access_policies::get_access_policy;
use crate::db::commands::_command::{delete_db_entry_by_id, gql_placeholder, command_boilerplate};
use crate::db::commands::_shared::increment_map_edits::increment_map_edits_if_valid;
use crate::db::general::permission_helpers::{assert_user_can_delete, is_user_creator_or_mod};
use crate::db::general::sign_in::jwt_utils::{resolve_jwt_to_user_info, get_user_info_from_gql_ctx};
use crate::db::node_child_links::get_node_child_links;
use crate::db::nodes::{get_node, is_root_node};
use crate::db::users::User;
use crate::utils::db::accessors::AccessorContext;
use rust_shared::utils::db::uuid::new_uuid_v4_as_b64;
use crate::utils::general::data_anchor::{DataAnchorFor1};

use super::_command::{set_db_entry_by_id_for_struct, NoExtras};
use super::delete_node::{delete_node, DeleteNodeInput};

wrap_slow_macros!{

#[derive(Default)] pub struct MutationShard_UnlinkNode;
#[Object] impl MutationShard_UnlinkNode {
	async fn unlink_node(&self, gql_ctx: &async_graphql::Context<'_>, input: UnlinkNodeInput) -> Result<UnlinkNodeResult, GQLError> {
		command_boilerplate!(gql_ctx, input, unlink_node);
    }
}

#[derive(InputObject, Deserialize)]
pub struct UnlinkNodeInput {
	pub mapID: Option<String>,
	pub parentID: String,
	pub childID: String,
}

#[derive(SimpleObject, Debug)]
pub struct UnlinkNodeResult {
	#[graphql(name = "_useTypenameFieldInstead")] __: String,
}

}

#[derive(Default)]
pub struct UnlinkNodeExtras {
	pub allow_orphaning: bool,
}

pub async fn unlink_node(ctx: &AccessorContext<'_>, user_info: &User, input: UnlinkNodeInput, extras: UnlinkNodeExtras) -> Result<UnlinkNodeResult, Error> {
	let UnlinkNodeInput { mapID, parentID, childID } = input;
	let result = UnlinkNodeResult { __: gql_placeholder() };
	
	let child_number_of_parents = get_node_child_links(ctx, None, Some(&childID)).await?.len();
	let parent_to_child_links = get_node_child_links(ctx, Some(&parentID), Some(&childID)).await?;
	ensure!(parent_to_child_links.len() == 1, "There should be 1 and only 1 link between parent and child. Link count:{}", parent_to_child_links.len());

	let old_data = get_node(ctx, &childID).await?;

	let base_text = format!("Cannot unlink node #{}, since ", old_data.id.as_str());
	ensure!(is_user_creator_or_mod(user_info, old_data.creator.as_str()), "{base_text}you are not its owner. (or a mod)");
	ensure!(extras.allow_orphaning || child_number_of_parents > 1, "{base_text}doing so would orphan it. Try deleting it instead.");
	ensure!(!is_root_node(ctx, &old_data).await?, "{base_text}it's the root-node of a map.");
	//ensure!(!IsNodeSubnode(oldData), "{baseText}it's a subnode. Try deleting it instead.");

	for link in parent_to_child_links {
		delete_db_entry_by_id(ctx, "nodeChildLinks".to_owned(), link.id.to_string()).await?;
	}

	increment_map_edits_if_valid(&ctx, mapID).await?;

	Ok(result)
}